examples/grammar-parser.cpp

```diff
@@ -0,0 +1,330 @@
+                while (*pos && *pos != '\r' && *pos != '\n') {
+                    pos++;
+                }
+            } else {
+                pos++;
+            }
+        }
+        return pos;
+    }
+
+    const char * parse_name(const char * src) {
+        const char * pos = src;
+        while (is_word_char(*pos)) {
+            pos++;
+        }
+        if (pos == src) {
+            throw std::runtime_error(std::string("expecting name at ") + src);
+        }
+        return pos;
+    }
+
+    std::pair<uint16_t, const char *> parse_char(const char * src) {
+        if (*src == '\\') {
+            char esc = src[1];
+            if (esc == 'x') {
+                int first = hex_to_int(src[2]);
+                if (first > -1) {
+                    int second = hex_to_int(src[3]);
+                    if (second > -1) {
+                        return std::make_pair((first << 4) + second, src + 4);
+                    }
+                }
+                throw std::runtime_error(std::string("expecting \\xNN at ") + src);
+            } else if (esc == '"' || esc == '[' || esc == ']') {
+                return std::make_pair(esc, src + 2);
+            } else if (esc == 'r') {
+                return std::make_pair('\r', src + 2);
+            } else if (esc == 'n') {
+                return std::make_pair('\n', src + 2);
+            } else if (esc == 't') {
+                return std::make_pair('\t', src + 2);
+            }
+            throw std::runtime_error(std::string("unknown escape at ") + src);
+        } else if (*src) {
+            return std::make_pair(*src, src + 1);
+        }
+        throw std::runtime_error("unexpected end of input");
+    }
+
+    const char * parse_alternates(
+            parse_state       & state,
+            const char        * src,
+            const std::string & rule_name,
+            uint16_t            rule_id,
+            bool                is_nested);
+
+    const char * parse_sequence(
+            parse_state           & state,
+            const char            * src,
+            const std::string     & rule_name,
+            std::vector<uint16_t> & outbuf,
+            bool                    is_nested) {
+        size_t out_start = outbuf.size();
+
+        // sequence size, will be replaced at end when known
+        outbuf.push_back(0);
+
+        size_t last_sym_start = outbuf.size();
+        const char * pos = src;
+        while (*pos) {
+            if (*pos == '"') { // literal string
+                pos++;
+                last_sym_start = outbuf.size();
+                while (*pos != '"') {
+                    auto char_pair = parse_char(pos);
+                         pos       = char_pair.second;
+
+                    // each char of a literal is encoded as a "range" of char - char
+                    outbuf.push_back(2);
+                    outbuf.push_back(char_pair.first);
+                    outbuf.push_back(char_pair.first);
```

Comment:
It seems like it could be possible pack two bytes into a single `uint16_t` value? `hi << 8 | low` in little endian order.

Reply:
There is also the issue here that since you're reading from `char` and not `unsigned char` or `uint8_t`, any byte over 0x7f might become sign extended as negative integers when cast into 16 bits.

Reply:
Yeah, my thinking was to eventually treat the char values as UTF-16 code units, but punted on Unicode support for now. 

And yes, `parse_char` does a conversion from `char` in the input string to `uint16_t`. I was maybe thinking to limit the BNF grammar to ascii and offer unicode escapes for higher values.

Reply:
In that case there is not much point to using UTF-16, I would go all the way to UTF-32.

UTF-16 still has characters spanning multiple code units.

Reply:
Yes, but at worst you'd have a surrogate pair for non-BMP. I felt like that was a good tradeoff for per-char space usage, but can go to UTF-32 if we want.